src/librustc_mir/build/expr/as_place.rs

```diff
@@ -6,13 +6,71 @@ use crate::build::{BlockAnd, BlockAndExtension, Builder};
 use crate::hair::*;
 use rustc::mir::interpret::{PanicInfo::BoundsCheck};
 use rustc::mir::*;
-use rustc::ty::{CanonicalUserTypeAnnotation, Variance};
+use rustc::ty::{CanonicalUserTypeAnnotation, Ty, Variance};
 
 use rustc_data_structures::indexed_vec::Idx;
 
+#[derive(Clone)]
+struct PlaceBuilder<'tcx> {
```

Comment:
Can you add a doc comment here explaining the *role* of the `PlaceBuilder`? I would say something like:

---

`PlaceBuilder` is used to create places during MIR construction. It allows you to "build up" a place by pushing more and more projections onto the end, and then convert the final set into a place using the `into_place` method.

This is used internally when building a place for an expression like `a.b.c`. The fields `b` and `c` can be progressively pushed onto the place builder that is created when converting `a`.

Reply:
I would also *consider* pulling into its own file, but maybe it's not worth it.